Fix missing deprecation warning for Tensor.nonzero().

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40187 Fix missing deprecation warning for Tensor.nonzero().**

There were two issues:
1) The hand-written definition included an ambiguous default, which made the deprecated signature not selected.  This didn't match the handwritten torch.nonzero, now they do.
2) A parsing bug for empty argument lists meant the signature wasn't being marked as deprecated.

Differential Revision: [D22118236](https://our.internmc.facebook.com/intern/diff/D22118236)

gchanan committed:
Fix missing deprecation warning for Tensor.nonzero().

There were two issues:
1) The hand-written definition included an ambiguous default, which made the deprecated signature not selected.  This didn't match the handwritten torch.nonzero, now they do.
2) A parsing bug for empty argument lists meant the signature wasn't being marked as deprecated.

gchanan commented:
CC @fmassa.

gchanan commented:
I'm not 100% sure about the python_arg_parser change because it seems like a bit of an unanticipated code path.

mruberry commented on test/test_torch.py:
```diff
@@ -10708,6 +10708,15 @@ def assert_tuple_empty(tup, dim):
+        with self.maybeWarnsRegex(UserWarning, "This overload of nonzero is deprecated"):
+            x.nonzero()
+
+        with self.maybeWarnsRegex(UserWarning, "This overload of nonzero is deprecated"):
```
Same test twice?

gchanan replied:
copy-past error, this is supposed to be `torch.nonzero(x)`, which worked previously.

mruberry commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -481,6 +481,7 @@ FunctionSignature::FunctionSignature(const std::string& fmt, int index)
       throw std::runtime_error("missing closing parenthesis: " + fmt);
     }
     if (offset == last_offset) {
```
This case seems weird. When could this reasonably occur? It seems like, if anything, this code should assert that offset != last_offset.

mruberry replied:
That is, this case only happens (from my reading) if last_offset points to a substring starting with ", ". That is a real possibility if the string is "(, ", for example. But it seems like such strings should be an error?

gchanan replied:
that's one case.  The other case is the `if (offset == std::string::npos) {` case, which would happen with an empty parameter list (last_offset will be equal to offset because last_offset is one past `(` and offset is `)`.

So, I think an assert would work if we fixed up the rest of the logic in some way, I was just a little hesitant to do that because I don't know why it was written this way in the first place...

mruberry replied:
Good point. On rereading I think in the case that `offset == std::string::npos` the code can assert that the`find(')'` returns a position in the string, and then you can check for `offset == last_offset` and short-circuit there. This should avoid the confusion strings like "(, ". I was also worried that setting last_offset like this would result in an invalid start of a substring pointer, but I checked and it's OK to take a substring starting from the null terminator. So the unchecked substrings starting from last_offset should be OK later.

gchanan committed:
Update on "Fix missing deprecation warning for Tensor.nonzero()."


There were two issues:
1) The hand-written definition included an ambiguous default, which made the deprecated signature not selected.  This didn't match the handwritten torch.nonzero, now they do.
2) A parsing bug for empty argument lists meant the signature wasn't being marked as deprecated.

mruberry reviewed (approved):
Cool!

gchanan committed:
Update on "Fix missing deprecation warning for Tensor.nonzero()."


There were two issues:
1) The hand-written definition included an ambiguous default, which made the deprecated signature not selected.  This didn't match the handwritten torch.nonzero, now they do.
2) A parsing bug for empty argument lists meant the signature wasn't being marked as deprecated.

Differential Revision: [D22118236](https://our.internmc.facebook.com/intern/diff/D22118236)
